File: cpp/client_test.cpp
#include "socket.hpp"
#include "p2p.hpp"
#include <iostream>
#include <cerrno>


int main(int argc, char** argv){

    using namespace socket_tool;
    using namespace p2p;

    Address server_addr;

    if( argc >= 3 ){
        char* ip = argv[0];
        int port = std::atoi(argv[1]);
        std::cout << "connecting to " << ip << ":" << port << "\n";
        server_addr = Address{ ip, port };
    }else{
        std::cout << "connecting to default 127.0.0.1:50020 \n";
        server_addr = Address{ "127.0.0.1", 50020 };
    }

    
    UDP_socket sock{ Address{ "0.0.0.0", 0 } };
    Address target_addr;
    while( 1 ){
        sleep(1);
        try{
            constexpr size_t buff_size = 1200;
            char buffer[buff_size] = {0};
            size_t len = buff_size;

            sock.send_to(buffer, 1, server_addr);
            auto recv_from_addr = sock.recv_from(buffer, &len);

            std::cout << "recv from server: " << buffer << '\n';

            if( buffer[0] == 'O' && buffer[1] == 'K' )
                continue;
            break;            

        }catch(Exception e){
            std::cerr << "[EXCEPTION] " << e.message << " code: " << e.code << " cerrno: " << strerror(errno) << "\n";
        }
    }

    



}

File: python/client_test.py
import socket
import json
import time
import p2p
import random


MY_ADDR = ("0.0.0.0", 0)
SERVER_ADDR = ("127.0.0.1", 50020)

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.bind(MY_ADDR)

target_addr: tuple = None
while True:
    time.sleep(1)
    try:

        sock.sendto(b'\x00', SERVER_ADDR)
        data, addr = sock.recvfrom(1200)

        if data == b'OK':
            continue

        data = json.loads(data)
        target_addr = ( data["ip"], data["port"] )
        break

    except Exception as e:
        print(f"[ERROR] {e}")


print(f"other peer is: {target_addr[0]}:{target_addr[1]}")

connection = p2p.P2PConnection(sock, target_addr[0], target_addr[1])

connection.punch_hole()

nums = {}
send_back = []
while 1:
    try:

        num = random.randrange(1, 100000)
        nums[num] = time.time()

        data = json.dumps({
            "echo": send_back,
            "send": num
        }).encode("utf-8")

        connection.send(data)
        data = connection.recv()

        data = json.loads(data)
        send_back = []
        send_back.append(data["send"])

        for echoed_num in data["echo"]:
            print( f"{round((time.time() - nums[echoed_num])*1000, 3)} ms -------> {echoed_num}" )
            del nums[echoed_num]


    except Exception as e:
        print(f"[ERROR] {e}")